Initial port of TD Tutorial

BlackstoneEngineering opened the pull request:
Original doc in this google doc : https://docs.google.com/document/d/12MgoVez96LvC8xwVO1DfjuMrPUYE-hUldr2tVRhS8fI/edit?usp=sharing 

@ARMmbed/mbed-docs  - please use the google doc as the guideline for what the tutorial should look like. I did my best to get the markdown right, but some things like embedded images / youtube videos I couldnt quite figure out.

BlackstoneEngineering committed:
Initial port of TD Tutorial

Original doc in this google doc : https://docs.google.com/document/d/12MgoVez96LvC8xwVO1DfjuMrPUYE-hUldr2tVRhS8fI/edit?usp=sharing 

@ARMmbed/mbed-docs  - please use the google doc as the guideline for what the tutorial should look like. I did my best to get the markdown right, but some things like embedded images / youtube videos I couldnt quite figure out.

iriark01 commented:
Amanda, I gave Austin the wrong path for his file. This should live with the other TD stuff in the Pelion Platform area. 

https://os.mbed.com/docs/mbed-os/v5.12/mbed-os-pelion/integrating-with-arm-treasure-data.html

Commit:
Make initial edits to td-tutorial.md

Make initial edits to file.

Commit:
Fix branding in td-tutorial.md

Fix branding.

AnotherButler commented on docs/tutorials/td-tutorial.md:
```diff
@@ -0,0 +1,226 @@
+
+## HTTPS library
+
+To use the HTTPS library, use this program: https://github.com/blackstoneengineering/mbed-os-example-treasuredata-rest. This program turns on Mbed OS device statistics by enabling the `MBED_ALL_STATS_ENABLED` macro and then send heap/CPU/stack/system information to Treasure Data.
```
Please move this example into an Arm repository.

BlackstoneEngineering replied:
No one has committed to long term maintinance of this library, so for the moment I would like to keep it in my personal repo space. 

@janjongboom - what do you think?

iriark01 replied:
Please move it to an Arm account. You can maintain it just as well from there.

AnotherButler commented on docs/tutorials/td-tutorial.md:
Are the two sections within this part of configuration, or should we change or remove this heading?

BlackstoneEngineering replied:
I dont understand the question, mind chatting with me about this monday? I cant seem to get the context correct in my head

AnotherButler replied:
Note to self: Remove h3 configure and replace with h3 setting Treasure Data databases and tables

AnotherButler commented on docs/tutorials/td-tutorial.md:
```diff
@@ -0,0 +1,226 @@
+
+For mass deployments. we recommend you use fluentd or fluentbit to aggregate and forward the data into Treasure Data. Depending on where you host your fluentd instance, you will need to follow slightly different setup instructions. (localhost on your machine with self signed certificates or at a public IP address in the cloud with Certificate Authority (CA) signed certificates). This example uses ...
+
+<INSERT YOUTUBE VIDEO FOR FLUENTD HERE: COMING SOON>
```
Query: When should we expect this?

BlackstoneEngineering replied:
Here : https://youtu.be/elB22i4y1yU

BlackstoneEngineering replied:
it has been completed.

BlackstoneEngineering replied:
still waiting for y'all to review the videos and give approval before I upload them to the armmbed channel

AnotherButler commented on docs/tutorials/td-tutorial.md:
````diff
@@ -0,0 +1,226 @@
+    
+   ```
+
+1. Wi-Fi credentials (optional): If you're using Wi-Fi, add your SSID/password. If you are using ethernet, you do not need to add Wi-Fi credentials.
````
Query: Is this actually optional? It sounds like it's not if you're using Wi-Fi. Also, do you add your SSID *and* password, or your SSID *or* password?

BlackstoneEngineering replied:
Its to imply that this works on Ethernet / Wifi / Cellular / etc. Its optional if you're not using wifi.

iriark01 replied:
I would remove (optional); I agree with Amanda that it's confusing.

AnotherButler commented on docs/tutorials/td-tutorial.md:
Query: What about Mbed Studio? We said above you can use it, but we don't show it here.

BlackstoneEngineering replied:
Mbed Studio is still beta software and does not fully support the F413ZH board yet, so im not including it. As soon as its stable and working I will start including it.

iriark01 replied:
Then why is it in the list?

AnotherButler reviewed (commented):
I've made some initial edits and left some initial queries for you to address.

Commit:
Make minor edits td-tutorial.md

Fix punctuation, and fix voice.

AnotherButler commented:
Please note I have not reviewed the videos yet.

megmiranda reviewed (commented):
I think Fleuntd should be init capped.

megmiranda commented on docs/tutorials/td-tutorial.md:
```diff
@@ -0,0 +1,224 @@
+There are two ways to send data securely from Mbed OS to Treasure Data:
+
+- HTTPS library - Send data directly to the Treasure Data REST API.
+- fluentd using fluent logger library - Send data to a hosted fluentd instance that aggregates and forwards the data on to your treasure data account.
```
ah ha this is how this works---I think Fluentd should be init capped

BlackstoneEngineering replied:
I was told by the inventor that its always lowercase, see the website : https://www.fluentd.org/ . 

I defer to your expertise.

megmiranda commented on docs/tutorials/td-tutorial.md:
```diff
@@ -0,0 +1,224 @@
+
+Both libraries are secured with Arm Mbed TLS in transit and are equally secure. We recommend the HTTPS library for development and the fluentd library for production. The tradeoff between the two is size of code on chip, size of data in transit and setup complexity: 
+
+- Code size on chip - The HTTPS library is ~50KB of ROM space on chip, this due to the HTTP stack. Both libraries use Mbed TLS to secure the connections, which is ~7KB per connection on your stack for both libraries.
```
on chip, this is due ....

megmiranda commented on docs/tutorials/td-tutorial.md:
```diff
@@ -0,0 +1,224 @@
+Both libraries are secured with Arm Mbed TLS in transit and are equally secure. We recommend the HTTPS library for development and the fluentd library for production. The tradeoff between the two is size of code on chip, size of data in transit and setup complexity: 
+
+- Code size on chip - The HTTPS library is ~50KB of ROM space on chip, this due to the HTTP stack. Both libraries use Mbed TLS to secure the connections, which is ~7KB per connection on your stack for both libraries.
+- Data size in transit - The HTTPS library sends data as a ASCII JSON string. The fluend library uses MessagePack (binary encoded json) across a TLS connection. This means that on average the fluentd library will use less bandwidth to send an equivalent message. When you pay per byte transmitted from both your power budget and data plan it matters.
```
Data size in transit - The HTTPS library sends data as an ASCII JSON string. The Fluend library uses MessagePack (binary encoded JSON) across a TLS connection. This means that on average the Fluentd library uses less bandwidth to send an equivalent message. When you pay per byte transmitted from both your power budget and data plan it matters.

megmiranda commented on docs/tutorials/td-tutorial.md:
```diff
@@ -0,0 +1,224 @@
+
+- Code size on chip - The HTTPS library is ~50KB of ROM space on chip, this due to the HTTP stack. Both libraries use Mbed TLS to secure the connections, which is ~7KB per connection on your stack for both libraries.
+- Data size in transit - The HTTPS library sends data as a ASCII JSON string. The fluend library uses MessagePack (binary encoded json) across a TLS connection. This means that on average the fluentd library will use less bandwidth to send an equivalent message. When you pay per byte transmitted from both your power budget and data plan it matters.
+- Maintenance - Initially, it may be simpler to setup the HTTPS library on a device and have it send data directly to treasure data, but what if you want to change what the device is doing or how its data is reported? If you are using the HTTPS library you will need to issue a firmware update to every device to change how it formats its data, but if you are using a fluend server you can simply mo ...
```
Maintenance - Initially, setting up the HTTPS library on a device and having it send data directly to Treasure Data is easier, but what if you want to change what the device is doing or how its data is reported? If you are using the HTTPS library you must issue a firmware update to every device to change how it formats its data. However, if you are using a Fluend server, you can modify the Fluentd config file on the server to change how data is formatted and processed.

megmiranda commented on docs/tutorials/td-tutorial.md:
```diff
@@ -0,0 +1,224 @@
+
+## HTTPS library
+
+To use the HTTPS library, use this program: https://github.com/blackstoneengineering/mbed-os-example-treasuredata-rest. This program turns on Mbed OS device statistics by enabling the `MBED_ALL_STATS_ENABLED` macro and then send heap/CPU/stack/system information to Treasure Data.
```
...and then sends heap.....

BlackstoneEngineering replied:
'and then sending heap....'

megmiranda commented on docs/tutorials/td-tutorial.md:
```diff
@@ -0,0 +1,224 @@
+
+### Import code
+
+You can compile the program through any of our three development tools:
```
You can compile the program using any of the following development tools:

megmiranda commented on docs/tutorials/td-tutorial.md:
```diff
@@ -0,0 +1,224 @@
+
+Next, you can compile and load your code onto your board. If you are unfamiliar with how to compile and load code, please look at the Mbed OS quick start tutorial.
+
+Once you have compiled your code and loaded it onto your board, open a serial terminal, and connect it to the board. View the output:
```
After you have...

BlackstoneEngineering replied:
?? not sure what this means?

AnotherButler replied:
I changed it to "After you have compiled your code ...."

megmiranda commented on docs/tutorials/td-tutorial.md:
```diff
@@ -0,0 +1,224 @@
+
+### Verify data in Treasure Data
+
+Go to the [Database list in Treasure data](https://console.treasuredata.com/app/databases), and open the `test_database` you created earlier. You can see the data from the board in the database. There is a 3- to 5-minute delay from when the data is sent to the database until the visualization system lets you see it, so please be patient, and wait for it to arrive. Be sure to refresh the page.
```
...Treasure Data ...

BlackstoneEngineering replied:
again, not sure what this comment means

AnotherButler replied:
I capitalized "Treasure Data" as requested by megmiranda.

megmiranda commented on docs/tutorials/td-tutorial.md:
```diff
@@ -0,0 +1,224 @@
+Now that you have data in Treasure data, it's time to analyze and use the data. 
+
+1. Go to the [Queries tab] (https://console.treasuredata.com/app/queries/editor).
+2. Select the `test_database`, and run some queries. To learn more about how to run queries, please read the [Treasure Data documentation](https://support.treasuredata.com/hc/en-us/articles/360001457427-Presto-Query-Engine-Introduction).
```
How about this article as an intro to running queries in TD .  https://support.treasuredata.com/hc/en-us/articles/360007995693

BlackstoneEngineering replied:
Sure, add that too. I took the article that was reccomended to me by the TD folks

megmiranda commented on docs/tutorials/td-tutorial.md:
```diff
@@ -0,0 +1,224 @@
+
+#### Select certain fields, order by time
+
+This query selects only certain columns from the table and then order them by the time field in ascending value, you can also replace `asc` with `desc` to get the order reversed.
```
....and orders them.... value. You can also ....

BlackstoneEngineering replied:
right, so dont make this in the comments, just make the change directly please.

megmiranda commented on docs/tutorials/td-tutorial.md:
...You need to rerun...

BlackstoneEngineering replied:
make the change

AnotherButler replied:
@BlackstoneEngineering , @megmiranda is our writer from Arm Treasure Data. I told her she could make the changes in comments, and I've already made these changes directly.

BlackstoneEngineering replied:
Understood. Its hard for me to tell what is being referenced as such a wide swath is highlighted in the github comment tool. I'll leave it to y'all

Commit:
Edit td-tutorial.md with changes from comments

Make changes requested in comments.

Commit:
Fix typo in td-tutorial.md

Fix typo.

AnotherButler commented on docs/tutorials/td-tutorial.md:
```diff
@@ -0,0 +1,224 @@
+
+## HTTPS library
+
+To use the HTTPS library, use this program: https://github.com/blackstoneengineering/mbed-os-example-treasuredata-rest. This program turns on Mbed OS device statistics by enabling the `MBED_ALL_STATS_ENABLED` macro and then sends heap/CPU/stack/system information to Treasure Data.
```
Please move this example to an Arm repository.

AnotherButler commented on docs/tutorials/td-tutorial.md:
```diff
@@ -0,0 +1,224 @@
+
+If you experience issues, ensure you have at least 10KB of space left on your stack. You can also change the `TD_DEBUG` macro to `true` to turn on the Treasure Data debug printfs.
+
+## fluentd
```
Query: Should this be capitalized?

BlackstoneEngineering replied:
should what be capitalized? The stuff in ` ` should be just as shown as they are copied directly from the config file. so `TD_DEBUG` should be exactly as it is.

AnotherButler commented on docs/tutorials/td-tutorial.md:
```diff
@@ -0,0 +1,224 @@
+There are two ways to send data securely from Mbed OS to Treasure Data:
+
+- HTTPS library - Send data directly to the Treasure Data REST API.
+- Fluentd using fluent logger library - Send data to a hosted Fluentd instance that aggregates and forwards the data on to your treasure data account.
```
What are the rules for capitalizing "fluent," "fluentd" and "fluend"?

megmiranda replied:
not super sure but i went to https://www.fluentd.org/ and it looked like it was init capped in text, lower case in logos.....kinda like Arm.

AnotherButler reviewed (commented):
I've left a few more queries.

Commit:
Fix capitalization in td-tutorial.md

Standardize capitalization of Fluentd.

AnotherButler commented:
This PR is failing Travis because the code snippets don't compile.

BlackstoneEngineering commented:
> This PR is failing Travis because the code snippets don't compile.

What code snippets? I have some quotes from the command line to show what success looks like, but i dont have any direct code snippets in here (i dont think)

AnotherButler commented:
@kegilbert Do you have time to look into this?

Commit:
Fix heading in td-tutorial.md

Fix heading to remove awkward double heading and make sense within the context of the document.

kegilbert commented on docs/tutorials/td-tutorial.md:
The code snippet compilation test is failing here since this is invalid JSON. I can't do multi-line suggestions, but this should work:

```suggestion
{
	"macros": [
		"MBEDTLS_SSL_DEBUG_ALL = 1"
	],
	"target_overrides": {
		"*": {
			"mbed-trace.enable": true
		}
	}
}
```

BlackstoneEngineering replied:
Ignore it and force publish. This needs to be there. This isnt a code snippet, its a quotation of what they can replace it with.

iriark01 replied:
Why not show it in a proper format?

BlackstoneEngineering replied:
because I dont want to copy the whole .json file, just the segment with things related to the debugging verbosity. This is to increase legibility.

iriark01 replied:
If you just want to give the name of two flags, then give the name of two flags. Why do you need a JSON code block for it? And if you want a JSON code block, then make it proper. 

Using a code block for something that cannot be copy/pasted as is is less legible than either of the other options.

BlackstoneEngineering replied:
Call it a stylistic difference of opinion.

kegilbert commented on docs/tutorials/td-tutorial.md:
If you take the above suggestion, remove this as well

```suggestion
```

BlackstoneEngineering replied:
again, ignore

BlackstoneEngineering committed:
moving youtube videos to Arm Mbed account post review

BlackstoneEngineering committed:
hiding the code names behind url links

BlackstoneEngineering commented:
@iriark01 / @AnotherButler - please make sure to include the images (per the google doc link example at the top), otherwise all changes have been made and it should be good to go, lets get it published!

iriark01 committed:
Update docs/tutorials/td-tutorial.md

Co-Authored-By: Kevin Gilbert <K.Gilbert.CH@gmail.com>

iriark01 committed:
Update docs/tutorials/td-tutorial.md

Co-Authored-By: Kevin Gilbert <K.Gilbert.CH@gmail.com>

Commit:
Update formatting in td-tutorial.md

Update video formatting for correct rendering.

iriark01 committed:
Adding links + file in JSON

iriark01 committed:
Adding images

iriark01 commented on docs/tutorials/td-tutorial.md:
```diff
@@ -0,0 +1,241 @@
+# Send data securely to Arm Treasure Data
+
+There are two ways to send data securely from Mbed OS to Treasure Data:
```
This is a very abrupt way to start. Also, what does "send from Mbed OS" mean? You can send from a device running Mbed OS.

BlackstoneEngineering replied:
+1

iriark01 committed:
That's not how image links work

iriark01 commented on docs/tutorials/td-tutorial.md:
```diff
@@ -0,0 +1,241 @@
+
+You can compile the program using any of the following development tools:
+
+- [Arm Mbed Studio](https://os.mbed.com/studio/).
```
Is there a way to give them a nice link?

BlackstoneEngineering replied:
Not sure what you mean by a nice link here? Like a direct import link from studio? Not to my knowledge. If you have a way please do so.

iriark01 commented on docs/tutorials/td-tutorial.md:
```diff
@@ -0,0 +1,241 @@
+You can compile the program using any of the following development tools:
+
+- [Arm Mbed Studio](https://os.mbed.com/studio/).
+- Arm Online Compiler - `ide.mbed.com/compiler?import=https://github.com/blackstoneengineering/mbed-os-example-treasuredata-rest`
```
Why isn't this a link, or the built in Import to Compiler option in the docs?

BlackstoneEngineering replied:
I'd love to have this be an import button that just takes them there. Please do so.

iriark01 commented on docs/tutorials/td-tutorial.md:
```diff
@@ -0,0 +1,241 @@
+- HTTPS library - Send data directly to the Treasure Data REST API.
+- Fluentd using fluent logger library - Send data to a hosted Fluentd instance that aggregates and forwards the data on to your treasure data account.
+
+Both libraries are secured with Arm Mbed TLS in transit and are equally secure. We recommend the HTTPS library for development and the Fluentd library for production. The tradeoff between the two is size of code on chip, size of data in transit and setup complexity:
```
"Both libraries are secured with Arm Mbed TLS in transit and are equally secure" is a bit clunky, as shown by using "secure" twice

BlackstoneEngineering replied:
Please feel free to reword it.

iriark01 commented on docs/tutorials/td-tutorial.md:
```diff
@@ -0,0 +1,241 @@
+
+- Code size on chip - The HTTPS library is ~50KB of ROM space on chip. This is due to the HTTP stack. Both libraries use Mbed TLS to secure the connections, which is ~7KB per connection on your stack for both libraries.
+- Data size in transit - The HTTPS library sends data as an ASCII JSON string. The Fluend library uses MessagePack (binary encoded JSON) across a TLS connection. This means that on average the Fluentd library uses less bandwidth to send an equivalent message. When you pay per byte transmitted from both your power budget and data plan it matters.
+- Maintenance - Initially, setting up the HTTPS library on a device and having it send data directly to Treasure Data is easier, but what if you want to change what the device is doing or how its data is reported? If you are using the HTTPS library, you must issue a firmware update to every device to change how it formats its data. However, if you are using a Fluend server, you can modify the Flu ...
```
I would say "whereas", not "however"; I don't think "however" is the right logical connection here

BlackstoneEngineering replied:
+1

iriark01 commented on docs/tutorials/td-tutorial.md:
```diff
@@ -0,0 +1,241 @@
+- Data size in transit - The HTTPS library sends data as an ASCII JSON string. The Fluend library uses MessagePack (binary encoded JSON) across a TLS connection. This means that on average the Fluentd library uses less bandwidth to send an equivalent message. When you pay per byte transmitted from both your power budget and data plan it matters.
+- Maintenance - Initially, setting up the HTTPS library on a device and having it send data directly to Treasure Data is easier, but what if you want to change what the device is doing or how its data is reported? If you are using the HTTPS library, you must issue a firmware update to every device to change how it formats its data. However, if you are using a Fluend server, you can modify the Flu ...
+
+The following steps show how to send data using first the HTTPS library and then using Fluentd.
```
Don't need the second "using" because of where the word "first" is in the sentence

BlackstoneEngineering replied:
+1

iriark01 commented on docs/tutorials/td-tutorial.md:
```diff
@@ -0,0 +1,241 @@
+- Arm Online Compiler - `ide.mbed.com/compiler?import=https://github.com/blackstoneengineering/mbed-os-example-treasuredata-rest`
+- Arm Mbed CLI (offline) - `mbed import https://github.com/blackstoneengineering/mbed-os-example-treasuredata-rest`
+
+### Setup variables
```
Set up

BlackstoneEngineering replied:
+1

iriark01 commented on docs/tutorials/td-tutorial.md:
```diff
@@ -0,0 +1,241 @@
+
+### Setup variables
+
+1. Configure the Treasure Data API key in `mbed_app.json` by change the `api-key` variable:
```
by changing

iriark01 commented on docs/tutorials/td-tutorial.md:
```diff
@@ -0,0 +1,241 @@
+
+### Compile and load
+
+Next, you can compile and load your code onto your board. If you are unfamiliar with how to compile and load code, please look at the Mbed OS quick start tutorial.
```
not familiar. Unfamiliar is something else.

iriark01 commented on docs/tutorials/td-tutorial.md:
```diff
@@ -0,0 +1,241 @@
+
+### Run queries
+
+Now that you have data in Treasure Data, it's time to analyze and use the data.
```
Can you say "analyze and use it"?

iriark01 commented on docs/tutorials/td-tutorial.md:
```diff
@@ -0,0 +1,241 @@
+
+#### Set configuration file
+
+Run Fluentd using the provided configuration file `fluentd --config ./fluentd-setup/fluentd.conf -vv`. This file opens two ports, port 24227 for unencrypted TCP traffic and port 24228 for TLS encrypted traffic. The configuration is provided for reference. We strongly suggest using TLS encryption on port 24228 to secure your data in transit.
```
This file opens two ports, 

Should that be : instead of ,

iriark01 commented on docs/tutorials/td-tutorial.md:
That's a very long sentence. I suggest breaking it down.

iriark01 commented on docs/tutorials/td-tutorial.md:
This is a little free of context unless you're already a TD user, I think

iriark01 replied:
This one I can't fix on my own, since I'm not a TD user. @megmiranda can you offer the readers more context?

iriark01 committed:
Higher quality images

BlackstoneEngineering commented:
@docs team, please consider all gramatical and minor changes pre-approved, feel free to re-write sentences as you see fit.

iriark01 committed:
Edits

iriark01 committed:
New videos
